### PR TITLE
Stabilize HaveHttpResponseMatcher

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Stabilize our custom HaveHttpResponseMatcher to avoid a panic occurring

# Context

https://github.com/solo-io/gloo/issues/8003#issuecomment-1553001321

I believe this change will stabilize our custom matcher. If all tests pass with this change, then I'm confident we can merge it. I do not know if this will address the flakey AWS test in the above github issue, but there are 2 possibilities:
- There is a true flake in that test, which will need to be addressed
- The test eventually passes, but because we encounter a panic when the first failure occurs, we do not retry. If this is the case, this fix should resolve the flake.

## Failure Message Examples
There were valid questions raised about the new message that would occur on failures, and whether it would contain the relevant details for developers to be able to understand and know how to fix. I modified our[ example test](https://github.com/solo-io/gloo/blob/main/test/e2e/example_test.go#L59), and ran it locally:
```
TEST_PKG=./test/e2e make test
```

and it produced:
```
[FAILED] Timed out after 5.000s.
  POST with request body should return same body in response
  The function passed to Eventually failed at /Users/samheilbron/go/src/github.com/solo-io/gloo/test/e2e/example_test.go:75 with:
  Expected
      <string>: some custom data
  to equal
      <string>: bad

  expected: {
    "StatusCode": 200,
    "Body": "bad",
    "Headers": null,
    "Custom": null
  }
  In [It] at: /Users/samheilbron/go/src/github.com/solo-io/gloo/test/e2e/example_test.go:76 @ 05/18/23 10:32:53.97
```

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/main/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: test/gomega/matchers/have_http_response.go [hunk 1]
```diff
@@ -86,45 +85,39 @@ func HaveHttpResponse(expected *HttpResponse) types.GomegaMatcher {
...
+		partialResponseMatchers = append(partialResponseMatchers, &matchers.HaveHTTPHeaderWithValueMatcher{
+			Header: headerName,
+			Value:  headerMatch,
+		})
+	}
+	partialResponseMatchers = append(partialResponseMatchers, expectedCustomMatcher)
+
+	return &HaveHttpResponseMatcher{
+		Expected:        expected,
+		responseMatcher: gomega.And(partialResponseMatchers...),
```
Comment (gunnar-solo) [inline test/gomega/matchers/have_http_response.go:113]: trying to understand, here.  Why is the `gomega.And` unguarded?  I thought the main panic was caused by sending in potentially non-errored objects to this method

File: test/gomega/matchers/have_http_response.go [hunk 2]
```diff
@@ -142,57 +135,21 @@ func (m *HaveHttpResponseMatcher) Match(actual interface{}) (success bool, err e
...
-		return false, customMatchErr
+	if ok, matchErr := m.responseMatcher.Match(actual); !ok {
+		return false, matchErr
 	}
 
 	return true, nil
 }
 
 func (m *HaveHttpResponseMatcher) FailureMessage(actual interface{}) (message string) {
-	return fmt.Sprintf("%s\n%s\n%s\n%s\n\ndiff: %s",
```
Comment (gunnar-solo) [inline test/gomega/matchers/have_http_response.go:167]: I'm trying to understand if this new error message format still can contain all of the previous information.

If not:  _can_ it do so?  Ex:  with some guarded string building?
If so:  help me understand what's going on?

File: test/gomega/matchers/have_http_response.go [hunk 3]
```diff
@@ -143,13 +144,23 @@ func (m *HaveHttpResponseMatcher) Match(actual interface{}) (success bool, err e
...
-		diff(m.Expected, actual))
+		informativeComparison(m.Expected, actual))
+}
+
+// informativeComparison returns a string which presents data to the user to help them understand why a failure occurred.
+// The HaveHttpResponseMatcher uses an And matcher, which intentionally short-circuits and only
+// logs the first failure that occurred.
+// To help developers, we print the entire actual/expected objects in case there were other inconsistencies
+func informativeComparison(expected, actual interface{}) string {
+	expectedJson, _ := json.MarshalIndent(expected, "", "  ")
```
Comment (ben-taussig-solo) [inline test/gomega/matchers/have_http_response.go:163]: can we format `actual` in the same way we format `expected`?

Comment (jbohanon) [reply test/gomega/matchers/have_http_response.go:165]: per huddle and other slack conversation, this in non-trivial. We know the `actual` will be of type `*http.Response`, but we are not able to guarantee that we can access the body

**Pull request merged**